File: langs/python/forgehx/src/forgehx/forgex.py
"""ForgeX sponge (SPECIFICATION_X §4)."""

from __future__ import annotations

import struct

from . import forgeperm

STATE_WORDS = 16
RATE_WORDS = 8
RATE_BYTES = RATE_WORDS * 8


class ForgeX:
    def __init__(self) -> None:
        self._state = [0] * STATE_WORDS
        self._offset = 0
        self._squeezing = False

    @staticmethod
    def hash(domain_tag: str, data: bytes) -> bytes:
        x = ForgeX()
        x.absorb_domain(domain_tag, data)
        return x.squeeze(32)

    @staticmethod
    def xof(domain_tag: str, data: bytes, length: int) -> bytes:
        if length < 0:
            raise ValueError("ForgeX: negative length")
        x = ForgeX()
        x.absorb_domain(domain_tag, data)
        return x.squeeze(length)

    def absorb_domain(self, domain_tag: str, data: bytes) -> None:
        tag = domain_tag.encode("ascii")
        self.absorb(struct.pack("<I", len(tag)))
        self.absorb(tag)
        self.absorb(data)

    def absorb(self, data: bytes) -> None:
        if self._squeezing:
            raise RuntimeError("ForgeX: cannot absorb after squeezing")
        rate = self._rate_bytes()
        for b in data:
            rate[self._offset] ^= b
            self._offset += 1
            if self._offset == RATE_BYTES:
                self._write_rate(rate)
                forgeperm.permute(self._state)
                self._offset = 0
                rate = self._rate_bytes()
        self._write_rate(rate)

    def squeeze(self, length: int) -> bytes:
        if not self._squeezing:
            self._pad_and_switch()
            self._squeezing = True
        out = bytearray(length)
        written = 0
        rate = self._rate_bytes()
        while written < length:
            if self._offset == RATE_BYTES:
                forgeperm.permute(self._state)
                self._offset = 0
                rate = self._rate_bytes()
            n = min(RATE_BYTES - self._offset, length - written)
            out[written : written + n] = rate[self._offset : self._offset + n]
            self._offset += n
            written += n
        return bytes(out)

    def _pad_and_switch(self) -> None:
        rate = self._rate_bytes()
        rate[self._offset] ^= 0x01
        rate[RATE_BYTES - 1] ^= 0x80
        self._write_rate(rate)
        forgeperm.permute(self._state)
        self._offset = 0

    def _rate_bytes(self) -> bytearray:
        return bytearray(struct.pack("<" + "Q" * RATE_WORDS, *self._state[:RATE_WORDS]))

    def _write_rate(self, rate: bytearray) -> None:
        words = struct.unpack("<" + "Q" * RATE_WORDS, bytes(rate))
        self._state[:RATE_WORDS] = list(words)

File: src/ForgeHash.X.Core/ForgeX.cs
using System.Buffers.Binary;
using System.Text;

namespace ForgeHashX;

/// <summary>
/// ForgeX sponge (SPECIFICATION_X §4). Experimental — not for production use.
/// </summary>
public sealed class ForgeX
{
    public const int StateWords = 16;
    public const int RateWords = 8;
    public const int RateBytes = RateWords * sizeof(ulong);

    private readonly ulong[] _state = new ulong[StateWords];
    private int _offset;
    private bool _squeezing;

    public static byte[] Hash(string domainTag, ReadOnlySpan<byte> data)
    {
        var x = new ForgeX();
        x.AbsorbDomain(domainTag, data);
        return x.Squeeze(32);
    }

    public static byte[] Xof(string domainTag, ReadOnlySpan<byte> data, int length)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(length);
        var x = new ForgeX();
        x.AbsorbDomain(domainTag, data);
        return x.Squeeze(length);
    }

    public void AbsorbDomain(string domainTag, ReadOnlySpan<byte> data)
    {
        byte[] tag = Encoding.ASCII.GetBytes(domainTag);
        Span<byte> len = stackalloc byte[4];
        BinaryPrimitives.WriteInt32LittleEndian(len, tag.Length);
        Absorb(len);
        Absorb(tag);
        Absorb(data);
    }

    public void Absorb(ReadOnlySpan<byte> input)
    {
        if (_squeezing)
        {
            throw new InvalidOperationException("Cannot absorb after squeezing.");
        }

        Span<byte> rate = RateAsBytes();
        foreach (byte b in input)
        {
            rate[_offset] ^= b;
            _offset++;
            if (_offset == RateBytes)
            {
                ForgePerm.Permute(_state);
                _offset = 0;
            }
        }
    }

    public byte[] Squeeze(int length)
    {
        if (!_squeezing)
        {
            PadAndSwitch();
            _squeezing = true;
        }

        byte[] output = new byte[length];
        int written = 0;
        Span<byte> rate = RateAsBytes();
        while (written < length)
        {
            if (_offset == RateBytes)
            {
                ForgePerm.Permute(_state);
                _offset = 0;
                rate = RateAsBytes();
            }

            int n = Math.Min(RateBytes - _offset, length - written);
            rate.Slice(_offset, n).CopyTo(output.AsSpan(written));
            _offset += n;
            written += n;
        }

        return output;
    }

    private void PadAndSwitch()
    {
        Span<byte> rate = RateAsBytes();
        rate[_offset] ^= 0x01;
        rate[RateBytes - 1] ^= 0x80;
        ForgePerm.Permute(_state);
        _offset = 0;
    }

    private Span<byte> RateAsBytes()
        => System.Runtime.InteropServices.MemoryMarshal.AsBytes(_state.AsSpan(0, RateWords));
}
